src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/ExternalSettingsProviderService.cs

```diff
@@ -0,0 +1,229 @@
+                            return Task.FromResult((ExternalSettingOperationResult)ExternalSettingOperationResult.Success.Instance);
+                        }
+                        break;
+                    }
+                case MonikerPackageRestoreAutomatic:
+                    {
+                        if (value is bool boolValue)
+                        {
+                            PackageRestoreConsent.IsAutomatic = boolValue;
+                            return Task.FromResult((ExternalSettingOperationResult)ExternalSettingOperationResult.Success.Instance);
+                        }
+                        break;
+                    }
+                case MonikerSkipBindingRedirects:
+                    {
+                        if (value is bool boolValue)
+                        {
+                            BindingRedirectBehavior.IsSkipped = boolValue;
+                            return Task.FromResult((ExternalSettingOperationResult)ExternalSettingOperationResult.Success.Instance);
+                        }
+                        break;
+                    }
+                case MonikerDefaultPackageManagementFormat:
+                    {
+                        if (value is string strValue)
+                        {
+                            PackageManagementFormat.SelectedPackageManagementFormat = strValue switch
+                            {
+                                MonikerPackagesConfig => 0,
+                                MonikerPackageReference => 1,
+                                _ => throw new ApplicationException("Error saving setting!"),
+                            };
+
+                            PackageManagementFormat.ApplyChanges();
+
+                            return Task.FromResult((ExternalSettingOperationResult)ExternalSettingOperationResult.Success.Instance);
+                        }
+                        break;
+                    }
+                case MonikerShowPackageManagementChooser:
+                    {
+                        if (value is bool boolValue)
+                        {
+                            PackageManagementFormat.Enabled = boolValue;
+                            PackageManagementFormat.ApplyChanges();
+                            return Task.FromResult((ExternalSettingOperationResult)ExternalSettingOperationResult.Success.Instance);
+                        }
+                        break;
+                    }
+                default: break;
+            }
+
+            throw new ApplicationException("Unknown setting!");
+        }
+
+        private static Task<ExternalSettingOperationResult<T>> ConvertValueOrThrow<T>(object input) where T : notnull
+        {
+            if (input is T value)
+            {
+                return Task.FromResult(ExternalSettingOperationResult.SuccessResult(value));
+            }
+
+            throw new ApplicationException("Error reading setting!");
```

Comment:
Would it perhaps be better to cast, so that the exception message tells us what the expected and actual types are, rather than a static string that we'll only be able to understand via debugging (perfview trace attached to "record your steps" VS feedback won't have details of the error)

Reply:
I've reworked the errors and described it in the new section, "Error handling". Please read that and see how this code looks. Thanks